### ISSUE TITLE
Wagtail Admin becomes extremely slow as the number of blocks increases.

### ISSUE THREAD

**Issue by mevinbabuc** [wagtail/wagtail#3062]
### Issue Summary

As the number of blocks (blocks with StreamField especially) increases on a _Page_, the admin becomes extremely slow. It becomes very visible after it has more than 10 blocks with _StreamField_ inside the body field of a _Page_. The page load time increases exponentially with every block you add after that.

With around 15 blocks in your StreamField, the page takes around 30 seconds of CPU time to finish processing it and the number of static files goes to 2600+

<img width="620" alt="screen shot 2016-10-09 at 7 40 57 am" src="https://cloud.githubusercontent.com/assets/1276650/19217300/202d8364-8df5-11e6-9411-f2f0cfeade94.png">

When looking close to the source of the rendered HTML, you can see these 2600+ scripts getting loaded or rendered on to the HTML, which takes the bulk of the processing time.

<img width="758" alt="screen shot 2016-10-09 at 7 54 15 am" src="https://cloud.githubusercontent.com/assets/1276650/19217322/c03cf416-8df5-11e6-974f-67cb6f0e5bb5.png">

On further inspecting the script tags, you get to know that most of the code or templates get repeated with a different id to the script tag.
### Questions
1. Is there a way to reduce this script tag inclusions ? or tell wagtail to reuse the scripts ?
2. Can we use template fragment caching, so that these scripts don't get re-evaluated on every page reload.
### Steps to Reproduce
1. Create a simple Page with a body field (StreamField) in it.
2. Add around 15 blocks to the body field (StreamField).
3. Try creating the Page in Wagtail Admin
### Technical details
- Python version: 3.4.
- Django version: 1.8.15.
- Wagtail version: 1.6.3
- Browser version: Chrome 53.

**Comment by mevinbabuc** [comment]
@gasman @m1kola Do you guys have any insight on this and any possible short term solutions for this ?

**Comment by gasman** [comment]
Hi @mevinbabuc,

Your "Steps to Reproduce" doesn't include enough information to reproduce this issue. If I test on a fresh Wagtail project with the following models.py, I get 15 `<script>` blocks, as I would expect:

    from __future__ import absolute_import, unicode_literals

    from django.db import models

    from wagtail.wagtailadmin.edit_handlers import StreamFieldPanel
    from wagtail.wagtailcore import blocks
    from wagtail.wagtailcore.fields import StreamField
    from wagtail.wagtailcore.models import Page


    class HomePage(Page):
        body = StreamField([
            ('text1', blocks.CharBlock()),
            ('text2', blocks.CharBlock()),
            ('text3', blocks.CharBlock()),
            ('text4', blocks.CharBlock()),
            ('text5', blocks.CharBlock()),
            ('text6', blocks.CharBlock()),
            ('text7', blocks.CharBlock()),
            ('text8', blocks.CharBlock()),
            ('text9', blocks.CharBlock()),
            ('text10', blocks.CharBlock()),
            ('text11', blocks.CharBlock()),
            ('text12', blocks.CharBlock()),
            ('text13', blocks.CharBlock()),
            ('text14', blocks.CharBlock()),
            ('text15', blocks.CharBlock()),
        ])

        content_panels = Page.content_panels + [
            StreamFieldPanel('body')
        ]

With more complex or deeply-nested blocks, the number of `<script>` blocks will naturally increase - for example, if the field contained 15 StreamBlock definitions with 15 sub-blocks in each one, then I would expect to see 15 * 15 = 225 `<script>` blocks. In this case, I wouldn't consider this a bug, as it's proportionate to the complexity of the block.

However, if you can provide an example of a models.py definition that generates far more `<script>` blocks than would be expected from the complexity of the block definition, I'll be happy to investigate further.

**Comment by mevinbabuc** [comment]
I'll try to put an example up, but we had 15 complex blocks with stream
fields inside them.

And I understand that script tag inclusion is proportional to the
complexity of the block, but is there anything that can be done to make it
not re-render them every time the page is loaded ?

Can't template fragment caching be used to make this faster someway ?

**Comment by gasman** [comment]
Template fragment caching might work... but as always with caching, the hard part is knowing when to clear the cache. For example, a ChoiceBlock with a dynamic list of choices (see #2812) would need to be rebuilt each time the list changed.

**Comment by kaedroho** [comment]
I don't think the issue is to do with the amount of Javascript blocks, or the block rendering itself. Django's templates should be able to handle both of these quickly at 15 blocks per page.

I think the problem is the way the block templates are being loaded: https://github.com/torchbox/wagtail/blob/master/wagtail/wagtailcore/blocks/field_block.py#L40

``render_to_string`` loads and compiles the template every time it is invoked, so in this case, it is loading and compiling the same template 15 times.

To make things worse, this is also happening on the field level: https://github.com/torchbox/wagtail/blob/master/wagtail/wagtailadmin/edit_handlers.py#L442

So if you have two fields in each block, this would load and compile 45 templates when it only really needs to do 3.

**Comment by gasman** [comment]
@kaedroho Interesting! I've tried tweaking the code to keep the compiled templates cached here: https://github.com/gasman/wagtail/commit/8f6e9a8c336b8e52036d447237ea864d5c75b745 - but it didn't seem to make much difference. Testing against a StreamField of 10 StreamBlocks of 10 CharBlocks each, it brought page render time down from ~3000ms to 2500ms. (I can open a PR, if you think it's worth it...)

**Comment by mevinbabuc** [comment]
I think any performance benefit is worth it, as it will lead to making the admin faster by few seconds when there are too many complex blocks. 

I think this particular tweak brings down loading time to 25 seconds from 30 seconds I'm experiencing now.

<img width="307" alt="screen shot 2016-10-13 at 8 25 18 am" src="https://cloud.githubusercontent.com/assets/1276650/19335153/ab176c34-911e-11e6-869f-a651cf4b1fb2.png">

**Comment by kaedroho** [comment]
Are there any other templates being loaded in this way? (comes to mind: widgets and edit handlers).

You could try mocking out a bit of Django's template loader and print out all the calls that are made.

You could also try running the cProfile module (which is how I discovered this originally) to find any remaining performance issues.

**Comment by jaap3** [comment]
The most simple solution to this is to recommend users to configure Django's cached template loader in production, right?

https://docs.djangoproject.com/en/1.10/ref/templates/api/#django.template.loaders.cached.Loader

**Comment by gasman** [comment]
PR for caching FieldBlock / StreamBlock templates added in #3077.

Will close this issue now, as we don't have a well-defined problem to solve here (e.g. we haven't identified a specific behaviour that's causing Wagtail to be substantially slower / more verbose than expected) - it boils down to "unusually complex StreamFields take unusually long to render", which isn't really fixable. Further optimisation work here is still welcome, of course.

**Comment by vsalvino** [comment]
I am seeing this issue as well. The more blocks a page contains the worse the performance. Pages with 20 or 30 blocks that contain 20 or 30 more nested blocks creates many tens of thousands of script tags and results in an HTML page that is several megabytes in size.

I think the ultimate solution would be to generate the contents of each script tag template dynamically via ajax when a block is added in the StreamField editor. This would provide O(1) performance as the number of script tag templates would no longer be related to the number of blocks. Only a small performance degradation would occur when clicking/selecting a block in the editor as the user would need to wait for an ajax call to complete to render the specified template.

This approach and usage of wagtail may go against its intended usage and design philosophies, but I plan to experiment with loading script templates on demand via ajax and will report back on this issue if anything comes of it. Would greatly appreciate any input on this approach from wagtail contributors.

**Comment by jellehermsen** [comment]
We have had the same issue at our company and we've started writing [a new StreamFieldPanel](https://github.com/UWKM/roadrunner) to alleviate the issue. Right now there's a very early alpha version that we have just started testing in some of our projects.

Our approach is quite simple:
- in our edit_handler we generate a tree with all the possible streamfields. Drilling down to the "primitive" Django fields as much as possible. We also generate lists of all the options for ChoiceFields and stick them in a HashMap, so we circumvent the duplication.
- this "StreamTree" is converted into JSON, and together with the choices map and the JSON from the latest page revision, we render this straight into a JS object
- We made an editor entirely in Javascript that figures out how to build all the HTML for the actual Panel. We embed HTML templates in the JS library, so we can do all the templating for the different streamfields clientside.
- Right now we try to stick to the original structure for the resulting POST data after you're done editing. Getting the paths for all the html inputs right is a bit tricky, so in the future we'll probably switch to client side validation for all inputs and simply generating the necessary JSON for the revision right there
- All in all this has resulted in a factor 30 increase in loading speed in some of our streamfield-heavy projects, which is quite a lot given the fact we were looking at 15 second edit page loads

Our project is very early right now and much of the code is pretty raw, however I do think this is the right direction. If the Wagtail devs think this approach has merit, I would love to work on it more and maybe get it into wagtail/contrib in the future.

**Comment by sect2k** [comment]
Like others, I've come across this issue. Like @mevinbabuc and @kaedroho  said, the one of the major causes of this is the excess rendering of form templates for adding new blocks, since the same template can get rendered multiple times.

This has multiple side effects, that together cause  `Page` Admin to become very slow to load.

1. Excessive calls to render template, in my case 2458 templates rendered according to `djdt`. This can be somewhat solved by using Django's cached template loader, but that's still around 2458 cache lookups.

2. Adding HTML for all those rendered forms increases size of download, again in my case it resulted in size of 1.8 MB, just for HTML. 

3. It also resulted in 345 `<script type="text/template">` tags to be added, which again has an effect on page loading time.

Some numbers for my page with a somewhat complex nested StreamField. This is on my dev machine.
```
Django Debug Toolbar:
Elapsed time:           16231.414
Request time:           16729
DOM loading time:       2956
Templates rendered:     2458                                    

Chrome Dev Tools:
HTML size:              1.8 MB
text/template tags:     345
Chrome load time:       20.19s
```

Since this was affecting one of my projects, I took a look at this today and have come up with a proof of concept, but working solution, that drastically reduces loading time for pages with complex StreamField implementations. It's implemented as a custom `StreamFieldPanel` and comes with it's own `StreamBlock` and `ListBlock` implementations. [Code that works with current release of wagtail is here](http://github.com/inueni/wagtail-smartstream).

Numbers after using `smartstream`.
```
Django Debug Toolbar:
Elapsed time:           7566.236                                
Request time:           7658                                    
DOM loading time:       2857                                                                     
Templates rendered:     355

Chrome Dev Tools:                                     
HTML size:              321 KB 
text/template tags:     19                                      
Chrome load time:       10.52s
```

As can be seen from the numbers, there is a 50% gain in load time, but what is most impressive is the reduction in number of `text/template' tags, down to just 19 from 345, a whooping 90% reduction, which also results in reduced HTML size, which decreased by 80%, from 1.8 MB down to 321 KB.

The solution is actually quite simple, not even 100 lines of python code, and only a couple of lines changed in two javascript files.

As is, Wagtail admin renders forms for adding blocks, by a chain of `all_html_declarations` and `html_declarations` calls, which ends with either a `StreamBlock` or `ListBlock` rendering a seperate template for each child block it has, and inserting a 'text/templates' with rendered template into `html_declarations` which get inserted into page.

To improved this, I moved the rendering and 'text/template' tag inserting process up the chain, in this case to `StreamFieldPanel`. This way it is possible to keep track of which block has already been rendered and inserted into page, so instead of getting multiple renders of the same block, it only renders it once and just stores a reference in JS object, which is later used to find the template for a specific block. Changes to wagtail stock js files (stream.js, list.js) are 4 lines in total.

As I mentioned this is just a proof of concept, but I've tested this and so far have not encountered any issues.

@gasman if you think this is something that could make it wagtail, I can prepare a PR, but before that, might be good to iron out the implementation details. Right now as I said, this works on StreamFieldPanel level, but ideally it should be even higher up the chain, probably `TabbedInterface`, to handle cases where there is more than one StreamField per page.

There is room for further improvement here, HTML for stream-menu is also included multiple times, where it could be rendered just once and then inserted on demand with JS.

**Comment by BertrandBordage** [comment]
To anyone reporting on this issue: did you try measuring response time without django-debug-toolbar?
Because this toolbar has an enormous cost on performance, I often saw pages respond 100× slower with the toolbar.

**Comment by gasman** [comment]
Reopening this to be the main ticket for tracking performance issues with nested StreamFields.

**Comment by mevinbabuc** [comment]
What I ended up doing was caching the rendered blocks to memcache and built
an API to fetch the blocks asynchronously rather than checking the html
template to pick the templates, which is currently done.. this has made the
whole admin extremely light weight and fast. The cached block templates are
rendered on demand, rather than rendering all at once.

**Comment by robertsoakes** [comment]
I'm curious, has there been any attempt to move this work into some of the streamfield replacements like `wagtail-react-streamfield`? It seems the solution is to move templating and edit logic to the frontend, and then delay rendering until a new block is actually invoked. That meshes well the stated goals of React Streamfield (and my understanding is that the react streamfield implementation has become a quasi-official project, with plans to incorporate it into the core).

My company has two projects that just are now effected by some of the nested slowness and would be interested in trying to implement some caching, frontend improvements, and optimization. @sect2k seems to have had some success by implementing cache logic, some of which might be migrated over to React Streamfield Are there any other efforts afoot?

(We don't want to step on alternative approaches.)

**Comment by robertsoakes** [comment]
After quite a lot of profiling in Wagtail React StreamField, we found that the slowdown for us is in the creation of the block signatures. The existing react frontend does a pretty good job of generating efficient HTML entities and pushing much of the templating logic to the browser.

I'm still working my way through the full source code to understand what is going on, but creating a stupidly simple cache system for block definitions speeds up the loading of the page by an order of magnitude. (Branch which shows the optimization can be found here: https://github.com/robertsoakes/wagtail-react-streamfield/tree/master/wagtail_react_streamfield.)

Are we going to start seeing any strange behavior from the pages? I still only have a limited understanding of all the pages that the cache definitions get used.

**Comment by robertsoakes** [comment]
I have submitted a pull request for an optional block cache:
https://github.com/wagtail/wagtail-react-streamfield/pull/42

This is pretty specific to our use-cases and sites, but ... we're seeing serious speed-ups on complex nested streamfields. After a recent update (which probably abused streamfield) we were seeing page loads of three or four minutes. After enabling the cache, we've managed to get that page load time down to three or four seconds (two orders of magnitude). In other (less dramatic cases), we've been able to reduce it down by more than half. Because the cache is not always desirable, we've put it behind a special edit handler.

I look forward to hearing comments and feedback :)

**Comment by pimarc** [comment]
I also came with this issue even after using wagtail-react-streamfield from Bertrand. In fact as soon as you have a complex nested streamfield structure, wagtail iterates all the tree for getting javascript static files and add them in any case. This may results in a very long loop (for me on a fast machine about 10 seconds and a loop of 26000 items). Sure, this occurs in admin edit only.
I investigated the code and found a very simple solution by implementing a kind of cache avoiding duplicates when searching for javascript files. The result is awesome in terms of speed.
The code is simple in the wagtail/core/blocks/base.py file. Just change the all_media method to the following one:

```    
    def all_media(self):
        media = forms.Media()
        media_cache = {}
        for block in self.all_blocks():
            key = block.media.__repr__()
            if key not in media_cache:
                media += block.media
                media_cache[key] = 1
        return media

```
I didn't find any issues for now with this modified code.

There is also another critiqual point with complex nested streamfields in the admin interface. The admin edit renders in html the complete tree in textarea with json value. In my case this text area is about 45Mb size with so many duplicate data !!!!
I think there is really something better to do here. I'm not a javascript specialist so I cannot really investigate on that. But I'm sure that something can be done here. So much duplicates in data is not a good thing.

**Comment by vsalvino** [comment]
For anyone looking for a quick way to reproduce this behavior, you can install `coderedcms` which has a noticeable streamfield slowness when running locally without template caching with sqlite.

1. `pip install coderedcms`
2. `coderedcms start myproject`
3. `python manage.py migrate`
4. `python manage.py createsuperuser`

Then runserver, log in to wagtail admin, and try to edit the Home Page. You will notice it takes several seconds to load (locally!) and produces over a megabyte of HTML.

You can see that our streamfield abuses the capabilities quite heavily, by having blocks containing blocks containing blocks... the streamfield definition on the default Home Page is outline here: https://github.com/coderedcorp/coderedcms/blob/e08c628e63ff56a462e056dc39c75542680a295b/coderedcms/blocks/__init__.py#L109

**Comment by thibaudcolas** [comment]
We’ve reviewed this during our team meeting and I believe @gasman plans to submit a PR with the suggested fix, even if this is just one aspect of the problem.

45Mb is quite a lot no matter how you cut it so I wonder if we’d need to consider having asynchronous rendering, taking the user’s position in the admin UI into account. But that’s for another day.

**Comment by pimarc** [comment]
Thank you very much Thibaud for taking care of my suggestion. After using the fix for several days it really reduces loading time by 5 in my case.

**Comment by thibaudcolas** [comment]
Credit goes to @zerolab for raising this during the meeting 🙂. Thanks for sharing this enhancement @pimarc – it definitely makes it easier for us to consider this and move forward.

**Comment by gasman** [comment]
This was one of the goals of the telepath StreamField rewrite - ensuring that when block definitions are nested, the generated HTML would scale linearly with the number of block definitions, rather than quadratically. With that released in 2.13, I think we can close this now.

**Comment by robicano22** [comment]
We are still having this issue, and I'm trying to figure out why...I upgraded from 2.11 to 2.13 - using pip3 from the CLI - this morning but am still lacking performance improvements.  A page still loads in the Admin with hundreds of script tags and takes more than 1 minute to get there.  I'm a bit newer to wagtail and Python/Django in general.  Let me know what else I should provide.  TIA
